crates/flox/src/config/mod.rs

```diff
@@ -326,6 +326,20 @@ impl Config {
 
         Ok(())
     }
+
+    /// If [Some] [Stability] provided, override config and set `FLOX_STABILITY` env var
+    ///
+    /// - The overriding stability is supplied by CLI args
+    /// - The env var will ensure that the selected is applied by default
+    ///   within a subshell started by flox
```

Comment:
Why is that desirable? That seems confusing

Reply:
I guess within a devshell you'd want to remain in the same stability context?
For activate we're not even actively using that and most other commands are not opening a subshell anyway.

This is here to _not_ introduce functional changes from before https://github.com/flox/flox/pull/215/files#diff-3bbe0f3376da952edd88493df9a55969a3b12f1e60f8c4bd0625fe19dc6ebc9aL196-L203 and before that https://github.com/flox/flox/blob/f94334c2bae1779044d339b5d22de1d06413e249/flox-bash/flox.sh#L47

Reply:
(it might be that the first use of this was just due to bash just using global variables a bit too generously)

Reply:
So would that mean that two flox builds would have different results?
```
flox build
flox develop --stability unstable
flox build
```
To me that seems undesirable. If you want to leave it in for the sake of preserving behavior, can you add a comment describing potentially undesirable cases?

Reply:
confirmed with @limeytexan that this is indeed the desired behavior

Reply:
No .. it was definitely intentional. `flox develop` in particular starts a subshell in which you would expect `flox build` to use the same stability by default. It's basically there because a) it shouldn't hurt anything to define FLOX_* variables in child processes and b) it propagates the selected stability to all child flox invocations.

However, speaking to Yannik on this topic he tells me that the mere presence of FLOX_STABILITY in the environment is _currently_ being used as an indication that flox should override the input (i.e. whether to operate in "locked" mode or not). _This is overloading the concept of stability_, and we instead need to have a separate control which determines whether to build/develop/publish/etc in locked mode or not.

Spelling out the concepts a little more in full:

* "stable" is compiled into the flox CLI itself as the default stability
* this can be overridden with the value of `$FLOX_STABILITY`
* a stronger override is to invoke `flox --stability <stability>`
* Each of the above sets the value of stability for a flox invocation, and that value is exported to the environment as `$FLOX_STABILITY` for the benefit of child flox invocations

Entirely separate from the above is the question of whether to run "locked" or "unlocked". I don't know what the UX for this will be, but _it is an entirely separate concern from the notion of stability_. My understanding is that Graham is also asking that people build "locked" by default.

With all that in mind, let's assume for argument' sake that we add a `flox -u` flag which implies one of "update" or "unlocked"; either way, the effect of running this way is to allow the chosen stability to move forward with each build.

By way of example, when building a capacitated (flox) flake:

* there are two relevant nixpkgs inputs:
    1. `floxpkgs/nixpkgs/nixpkgs`: is always overridden to refer to `nixpkgs-<stability>`
    2. `nixpkgs-<stability>`: the lock for whatever is the selected stability for that invocation
        * (bootstrap) it's worth noting flox will automatically lock this to the current stability at the time when first used, whether or not it's running in `-u` mode
* building in regular (locked) mode
    * the `floxpkgs/nixpkgs/nixpkgs` input will be overridden to refer to `nixpkgs-<stability>`
    * the `nixpkgs-<stability>` input will come from the lockfile
        * ... unless of course it's never been locked before, in which case it will automatically lock to the current stability
* building in unlocked mode
    * the `floxpkgs/nixpkgs/nixpkgs` input will be overridden to refer to `nixpkgs-<stability>`
    * the `nixpkgs-<stability>` input will be overridden to the current stability

... and that's it. It's no more complicated than that. Notably, when building a regular (non-capacitated) flake _the procedure is exactly the same_[^1], because a regular flake will use neither of the `floxpkgs/nixpkgs/nixpkgs` nor `nixpkgs-<stability>` inputs anyway, and will _always_ operate in "locked" mode as a result.

[^1]: This isn't _exactly_ true because we'll likely want to scan the flake to look for the existence of the `floxpkgs/nixpkgs/nixpkgs` input in order to avoid adding the `--override-input floxpkgs/nixpkgs` args and avoid spitting warnings to stderr, but that's a tangential issue.

Reply:
I'm grinding my teeth on *`floxpkgs/nixpkgs/nixpkgs`: is always overridden to refer to `nixpkgs-<stability>`* and *the `nixpkgs-<stability>` input will come from the lockfile*.

Afaik this is an eval performance issue and will also prevent automatic updates.
Its good however that we point out that _whether we apply_ stabilities should be separated from _which_ stability to apply. 

But this comment thread here is not the right place to discuss the technical details.

Reply:
> I'm grinding my teeth on floxpkgs/nixpkgs/nixpkgs: is always overridden to refer to nixpkgs-<stability> and the nixpkgs-<stability> input will come from the lockfile.

Heh .. I was quoting what I heard you say about how it worked.  :)

Bottom line: the `nixpkgs-<stability>` input gets locked, and then there is [a level of indirection](https://www2.dmst.aueb.gr/dds/pubs/inbook/beautiful_code/html/Spi07g.html#:~:text=All%20problems%20in%20computer%20science,envisioned%20the%20modern%20personal%20computer.) (however you want to do it) that selects which of those locked `nixpkgs-<stability>` inputs get used for a particular (capacitated) flake build. You handle the implementation details in whatever way makes sense.

Reply:
My understanding was that in "locked" mode there will be no override of any kind but we're rather doing a pure eval ( + whatever comes after ) based on the **unmodified** flake.lock.

Reply:
moving this discussion to https://github.com/flox/product/issues/448

Reply:
Even in locked mode you might want to try building with a different stability, e.g. `flox build --stability unstable` to see if the latest version fixes a problem. That's very common actually.